### ISSUE TITLE
bug: inconsistent behaviour with hooked configmap

### ISSUE THREAD

**Issue by mabdh** [odpf/charts#69]
**Description**
With helm version `v3.5.3`, used by terraform helm provider v2.1.2.

Helm hook annotation are added in configmap and secret in this [commit](https://github.com/odpf/charts/commit/fdeb281cda76efb85e8ce6ed34dbf79299156d23). During helm upgrade, the migration job pods are terminated along with the deletion of configmap and secret resources.

However, running helm upgrade multiple times cause different final states of upgrade.

Behaviour:
**1st upgrade**
Initial state:
- Config exist
- App pod exist

Hook running:
- Config removed
- Config added
- Migration pod running
- Migration terminated
- Config removed

Final state:
- Config not exist
- Old pods are running
- Some new pods are created with `CreateContainerConfigError`

**2nd upgrade**
Initial state:
- Config not exist
- Old app pods are running, New app pods are created with `CreateContainerConfigError`

Hook running:
- Config removed
- Config added
- Migration pod running
- Migration terminated
- Config removed

Final state:
- Config not exist
- Old pods are terminated
- Some new app pods are running, the other new app pods are created with `CreateContainerConfigError`

**3rd upgrade**
Initial state:
- Config not exist
- Some new app pods are running, some new app pods are created with `CreateContainerConfigError`

Hook running:
- Config removed
- Config added
- Migration pod running
- Migration terminated
- Config removed

Final state:
- Config not exist
- All new app pods are running

**4th upgrade**
Initial state:
- Config not exist
- All new app pods are running

Hook running:
- Config removed
- Config added
- Migration pod running
- Migration terminated
- Config removed

Final state:
- Config exist
- All new app pods are running (no-op)

**To Reproduce**
Run helm upgrade with odpf app chart version ` v0.4.0`

**Expected behavior**
Regardless failed or success, the migration jobs pods and configmap should not be removed when deployment is finished

**Comment by mabdh** [comment]
This is related with https://github.com/helm/helm/issues/8694

**Comment by mabdh** [comment]
So in our current state, we have
- configmap & secret
- App pod
- Migration pod (completed)

We are trying to redeploying again and it was running smoothly. I wonder if the existence of a migration in the initial state pod does matter when we have a hook. Because the issue that was mentioned in this thread (above), we do upgrade without having `migration pod` in our system.

**Comment by bsushmith** [comment]
We had the exact same issue while upgrading `odpf/guardian` at our end which included a migration. Even after multiple runs, the deleted `configmap` & `secrets` were not restored again.

So we manually created & applied `configmap` and `secrets` using `kubectl apply`. But when triggered a helm upgrade again, the manually created `configmap` and `secrets` were also removed. 

Then we added the annotation, for both `cm` and `secrets`, manually - `"helm.sh/resource-policy": keep`. 

Now, after this - on rerun of helm upgrade, the `cm` and `secrets` files were overwritten with what's coming through the pipeline but they didn't get deleted and stayed. After this, we ran multiple times again - but the configs didn't get deleted. 

If we add this annotation to our app chart, it should fix this annoying bug.


@snorlacs @rvirdiz
